File: Day-3/solution.php
<?php

$data = file_get_contents("input.txt");

printf("Part 1: %s\n", part1($data));
printf("Part 2: %s\n", part2($data));

function part1(string $data): int{
    $result = 0;
    $data = parseInput($data);

    $re = '/mul\((\d{1,3})\,(\d{1,3})\)/m';

    preg_match_all($re, $data, $matches, PREG_SET_ORDER, 0);

    foreach ($matches as $match) {
        $result += intval($match[1]) * intval($match[2]);
    }

    return $result;
}

function part2(string $data): int{
    $result = 0;
    $data = parseInput($data);

    $re = '/mul\((\d{1,3})\,(\d{1,3})\)|do\(\)|don\'t\(\)/m';

    preg_match_all($re, $data, $matches, PREG_SET_ORDER, 0);
    $do = true;
    foreach ($matches as $match) {

        if ($match[0] == "do()"){
            $do = true;
            continue;
        } else if ($match[0] == "don't()"){
            $do = false;
            continue;
        }

        if ($do && ($match[1])) {
            $result += intval($match[1]) * intval($match[2]);
        }
        
    }

    return $result;
}

function parseInput(string $data):string{
    return $data;
}

File: Day-3/solution.py
import re

def part1(data:str) -> int:
    result = 0
    regex = r"mul\((\d{1,3})\,(\d{1,3})\)"

    matches = re.finditer(regex, data, re.MULTILINE)
    for _, match in enumerate(matches, start=1):
        left, right = match.group()[4:-1].split(',')
        result = result + (int(left) * int(right))
    
    return result
        

def part2(data:str) -> int:
    result = 0
    regex = r"mul\((\d{1,3})\,(\d{1,3})\)|do\(\)|don\'t\(\)"

    matches = re.finditer(regex, data, re.MULTILINE)
    do = True

    for _, match in enumerate(matches, start=1):
        if match.group() == "do()":
            do = True
            continue
        elif match.group() == "don't()":
            do = False
            continue
        if do:
            left, right = match.group()[4:-1].split(',')
            result = result + (int(left) * int(right))

    return result
        

if __name__ == "__main__":
    
    with open("input.txt") as f:
        temp = f.read()
        
        
    print(f"Part 1: {part1(temp)}")
    print(f"Part 2: {part2(temp)}")
